assets/pages/regular-contributions/components/formFields.jsx

```diff
@@ -127,27 +122,48 @@ function countriesDropdown(
 
 function NameForm(props: PropTypes) {
 
+  const showFirstNameError = props.firstName.shouldValidate && !props.firstName.isValid(props.firstName.value);
+  const showLastNameError = props.lastName.shouldValidate && !props.lastName.isValid(props.lastName.value);
+  const firstNameModifier = showFirstNameError
+    ? ['first-name', 'error']
+    : ['first-name'];
+  const lastNameModifier = showLastNameError
+    ? ['last-name', 'error']
+    : ['last-name'];
+
   return (
     <form className="regular-contrib__name-form">
-      <EmailFormFieldContainer />
+      <EmailFormFieldContainer
+        dispatch={props.dispatch}
+      />
       <TextInput
         id="first-name"
         labelText="First name"
         placeholder="First name"
-        value={props.firstName}
-        onChange={props.firstNameUpdate}
-        modifierClasses={['first-name']}
+        value={props.firstName.value}
+        onChange={props.firstName.setValue(props.dispatch)}
```

Comment:
for `INPUT:any([type="text"],[type="email"],[type="password"],[type="tel'],[type="url"])`, I'd use the `onInput` event, whereas `onChange` is better suited for `SELECT` elements.

Reply:
onChange and onInput are identical in React TextInput https://stackoverflow.com/questions/38256332/in-react-whats-the-difference-between-onchange-and-oninput

Reply:
oh, had no idea, thanks for the link 👍